File: Mac/ES_R_DevKit_Mac/MEME_Academic/SettingViewController.m
//
//  SettingViewController.m
//  MEME_Academic
//
//  Created by Celleus on 2022/09/02.
//  Copyright © 2022 jins-jp. All rights reserved.
//

#import "SettingViewController.h"
#import "Const.h"
#import "Common.h"
#import "UserSetting.h"

@interface SettingViewController ()

@property (weak) IBOutlet NSTextField *browseTextField;

@property (weak) IBOutlet NSTextFieldCell *xAxisTextField;
@property (weak) IBOutlet NSTextFieldCell *yAxisTextField;
@property (weak) IBOutlet NSTextFieldCell *zAxisTextField;

@property (weak) IBOutlet NSButton *showSaveFileDialogButton;
@property (weak) IBOutlet NSButton *extermalOutputSocketButton;

@property (weak) IBOutlet NSTextFieldCell *localProtTextField;
@property (weak) IBOutlet NSTextField *localAddressTextField;

@end

@implementation SettingViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do view setup here.
    
    [self setSettingParameter];
    // ボタンのスタイルを設定（ライト／ダークモード対応）
    [self styleButtonsForAppearance];
}

- (void)viewWillAppear {
    [super viewWillAppear];
    self.view.window.title = @"Setting";
}

- (void)setSettingParameter {
    NSLog(@"setSettingParameter");
    _browseTextField.stringValue = [UserSetting getSaveFilePath];
    _xAxisTextField.stringValue = [NSString stringWithFormat:@"%g",[UserSetting getXAxis]];
    _yAxisTextField.stringValue = [NSString stringWithFormat:@"%g",[UserSetting getYAxis]];
    _zAxisTextField.stringValue = [NSString stringWithFormat:@"%g",[UserSetting getZAxis]];
    _showSaveFileDialogButton.state = [UserSetting getShowSaveFileDialog];
    _extermalOutputSocketButton.state = [UserSetting getExtermalOutputSocket];
    _localProtTextField.stringValue = [UserSetting getLocalPort];
    _localAddressTextField.stringValue = [Common getIPAddress];
}

- (void)saveSettingParameter {
    NSLog(@"saveSettingParameter");
    [UserSetting setSaveFilePath:_browseTextField.stringValue];
    [UserSetting setXAxis:[_xAxisTextField.stringValue doubleValue]];
    [UserSetting setYAxis:[_yAxisTextField.stringValue doubleValue]];
    [UserSetting setZAxis:[_zAxisTextField.stringValue doubleValue]];
    [UserSetting setShowSaveFileDialog:_showSaveFileDialogButton.state];
    [UserSetting setExtermalOutputSocket:_extermalOutputSocketButton.state];
    [UserSetting setLocalPort:_localProtTextField.stringValue];
//    [UserSetting setLocalAddress:_localAddressTextField.stringValue];
}

// =============================================================================
#pragma mark - button_StartScan_Tapped
// =============================================================================
- (IBAction)button_Browse_Tapped:(id)sender {
    NSLog(@"button_Browse_Tapped");
    [self showOpenPanel];
}

- (void)showOpenPanel {
    NSOpenPanel *panel = [[NSOpenPanel alloc] init];
    panel.canChooseDirectories = true;
    panel.canChooseFiles = false;
    [panel beginSheetModalForWindow:self.view.window
                  completionHandler:^(NSModalResponse result) {
        
        switch (result) {
            case NSModalResponseOK:
                _browseTextField.stringValue = panel.URL.path;
                break;
            default:
                break;
        }
        
    }];
}

// =============================================================================
#pragma mark - button_OpenFolder_Tapped
// =============================================================================
- (IBAction)button_OpenFolder_Tapped:(id)sender {
    NSLog(@"button_OpenFolder_Tapped");
    [self showOpenFolder];
}

- (void)showOpenFolder {
    NSString *urlString = _browseTextField.stringValue;
    NSURL *url = [NSURL fileURLWithPath:urlString];
    if ([[NSWorkspace sharedWorkspace] openURL:url]) {
        NSLog(@"ファイルが開ける");
    }
    else {
        NSLog(@"ファイルが開けない");
    }
}

// =============================================================================
#pragma mark - button_Extemal_Output_Socket_Tapped
// =============================================================================
- (IBAction)button_Extemal_Output_Socket_Tapped:(id)sender {
    NSLog(@"button_Extemal_Output_Socket_Tapped");
}

// =============================================================================
#pragma mark - textField_Local_Port_Tapped
// =============================================================================
- (IBAction)textField_Local_Port_Tapped:(id)sender {
    NSLog(@"textField_Local_Port_Tapped");
}

// =============================================================================
#pragma mark - button_Apply_Tapped
// =============================================================================
- (IBAction)button_Apply_Tapped:(id)sender {
    NSLog(@"button_Apply_Tapped");
    [self saveSettingParameter];
    if (self.delegate) {
        [self.delegate didApply:self];
    }
    [self.view.window close];
}

// =============================================================================
#pragma mark - button_Cancel_Tapped
// =============================================================================
- (IBAction)button_Cancel_Tapped:(id)sender {
    NSLog(@"button_Cancel_Tapped");
    [self.view.window close];
}

- (void)styleButtonsForAppearance {
    NSString *appearanceName = [self.view.effectiveAppearance bestMatchFromAppearancesWithNames:@[NSAppearanceNameAqua, NSAppearanceNameDarkAqua]];
    BOOL isDark = [appearanceName isEqualToString:NSAppearanceNameDarkAqua];
    CGFloat bgWhite = isDark ? 0.60 : 0.85;
    NSColor *titleColor = isDark ? [NSColor labelColor] : [NSColor colorWithWhite:0.15 alpha:1.0];
    // Browse/Open/Apply/Cancel should keep the same (dark) title color even in dark mode
    NSColor *constantButtonTitleColor = [NSColor colorWithWhite:0.15 alpha:1.0];
    [self applyStyleToButtonsInView:self.view bgWhite:bgWhite titleColor:titleColor];
}

- (void)applyStyleToButtonsInView:(NSView *)parent bgWhite:(CGFloat)bg titleColor:(NSColor *)titleColor {
    for (NSView *sub in parent.subviews) {
        if ([sub isKindOfClass:[NSButton class]]) {
            NSButton *button = (NSButton *)sub;
            // 除外: Show save file dialog と External output socket のボタンは変更しない
            if (button == _showSaveFileDialogButton || button == _extermalOutputSocketButton) {
                // skip styling for these specific controls
            } else {
                button.wantsLayer = YES;
                button.layer.cornerRadius = 14.0;
                button.layer.masksToBounds = YES;
                button.layer.backgroundColor = [NSColor colorWithWhite:bg alpha:1.0].CGColor;
                // 枠無しにする
                button.bordered = NO;
                if ([button respondsToSelector:@selector(cell)]) {
                    id cell = button.cell;
                    if ([cell respondsToSelector:@selector(setBordered:)]) {
                        [cell setBordered:NO];
                    }
                }
                // Decide title color: keep constant for Browse/Open/Apply/Cancel, otherwise use appearance-based color
                NSColor *useTitleColor = titleColor;
                if (button == _button_Browse || button == _button_OpenFolder || button == _button_Apply || button == _button_Cancel) {
                    useTitleColor = [NSColor colorWithWhite:0.15 alpha:1.0];
                }

                NSMutableAttributedString *attr = [[NSMutableAttributedString alloc] initWithAttributedString:button.attributedTitle];
                NSRange range = NSMakeRange(0, attr.length);
                [attr addAttribute:NSForegroundColorAttributeName value:useTitleColor range:range];
                [button setAttributedTitle:attr];
            }
        }
        [self applyStyleToButtonsInView:sub bgWhite:bg titleColor:titleColor];
    }
}

@end

File: Mac/ES_R_DevKit_Mac/MEME_Academic/SettingViewController.swift
//
//  SettingViewController.swift
//  MEME_Academic
//
//  Created by Celleus on 2022/09/02.
//  Copyright © 2022 jins-jp. All rights reserved.
//

import Cocoa

@MainActor
protocol SettingViewControllerDelegate: AnyObject {
    func didApply(_ settingViewController: SettingViewController)
}

class SettingViewController: NSViewController {

    weak var delegate: SettingViewControllerDelegate?

    @IBOutlet weak var button_Browse: NSButton!
    @IBOutlet weak var button_OpenFolder: NSButton!
    @IBOutlet weak var button_Apply: NSButton!
    @IBOutlet weak var button_Cancel: NSButton!

    @IBOutlet weak var browseTextField: NSTextField!

    @IBOutlet weak var xAxisTextField: NSTextFieldCell!
    @IBOutlet weak var yAxisTextField: NSTextFieldCell!
    @IBOutlet weak var zAxisTextField: NSTextFieldCell!

    @IBOutlet weak var showSaveFileDialogButton: NSButton!
    @IBOutlet weak var extermalOutputSocketButton: NSButton!

    @IBOutlet weak var localProtTextField: NSTextFieldCell!
    @IBOutlet weak var localAddressTextField: NSTextField!

    override func viewDidLoad() {
        super.viewDidLoad()
        setSettingParameter()
        styleButtonsForAppearance()
    }

    override func viewWillAppear() {
        super.viewWillAppear()
        view.window?.title = "Setting"
    }

    func setSettingParameter() {
        NSLog("setSettingParameter")
        browseTextField.stringValue = UserSetting.getSaveFilePath()
        xAxisTextField.stringValue = String(format: "%g", UserSetting.getXAxis())
        yAxisTextField.stringValue = String(format: "%g", UserSetting.getYAxis())
        zAxisTextField.stringValue = String(format: "%g", UserSetting.getZAxis())
        showSaveFileDialogButton.state = UserSetting.getShowSaveFileDialog() ? .on : .off
        extermalOutputSocketButton.state = UserSetting.getExtermalOutputSocket() ? .on : .off
        localProtTextField.stringValue = UserSetting.getLocalPort()
        localAddressTextField.stringValue = Common.getIPAddress()
    }

    func saveSettingParameter() {
        NSLog("saveSettingParameter")
        UserSetting.setSaveFilePath(browseTextField.stringValue)
        UserSetting.setXAxis(Double(xAxisTextField.stringValue) ?? 0)
        UserSetting.setYAxis(Double(yAxisTextField.stringValue) ?? 0)
        UserSetting.setZAxis(Double(zAxisTextField.stringValue) ?? 0)
        UserSetting.setShowSaveFileDialog(showSaveFileDialogButton.state == .on)
        UserSetting.setExtermalOutputSocket(extermalOutputSocketButton.state == .on)
        UserSetting.setLocalPort(localProtTextField.stringValue)
    }

    @IBAction func button_Browse_Tapped(_ sender: Any) {
        NSLog("button_Browse_Tapped")
        showOpenPanel()
    }

    private func showOpenPanel() {
        let panel = NSOpenPanel()
        panel.canChooseDirectories = true
        panel.canChooseFiles = false
        guard let window = view.window else { return }
        panel.beginSheetModal(for: window) { [weak self] result in
            if result == .OK, let url = panel.url {
                self?.browseTextField.stringValue = url.path
            }
        }
    }

    @IBAction func button_OpenFolder_Tapped(_ sender: Any) {
        NSLog("button_OpenFolder_Tapped")
        showOpenFolder()
    }

    private func showOpenFolder() {
        let urlString = browseTextField.stringValue
        let url = URL(fileURLWithPath: urlString)
        if NSWorkspace.shared.open(url) {
            NSLog("ファイルが開ける")
        } else {
            NSLog("ファイルが開けない")
        }
    }

    @IBAction func button_Extemal_Output_Socket_Tapped(_ sender: Any) {
        NSLog("button_Extemal_Output_Socket_Tapped")
    }

    @IBAction func textField_Local_Port_Tapped(_ sender: Any) {
        NSLog("textField_Local_Port_Tapped")
    }

    @IBAction func button_Apply_Tapped(_ sender: Any) {
        NSLog("button_Apply_Tapped")
        saveSettingParameter()
        delegate?.didApply(self)
        view.window?.close()
    }

    @IBAction func button_Cancel_Tapped(_ sender: Any) {
        NSLog("button_Cancel_Tapped")
        view.window?.close()
    }

    private func styleButtonsForAppearance() {
        let appearanceName = view.effectiveAppearance.bestMatch(from: [.aqua, .darkAqua])
        let isDark = appearanceName == .darkAqua
        let bgWhite: CGFloat = isDark ? 0.60 : 0.85
        let titleColor: NSColor = isDark ? .labelColor : NSColor(white: 0.15, alpha: 1.0)
        applyStyle(toButtonsIn: view, bgWhite: bgWhite, titleColor: titleColor)
    }

    private func applyStyle(toButtonsIn parent: NSView, bgWhite: CGFloat, titleColor: NSColor) {
        for sub in parent.subviews {
            if let button = sub as? NSButton {
                if button == showSaveFileDialogButton || button == extermalOutputSocketButton {
                    // skip
                } else {
                    button.wantsLayer = true
                    button.layer?.cornerRadius = 14.0
                    button.layer?.masksToBounds = true
                    button.layer?.backgroundColor = NSColor(white: bgWhite, alpha: 1.0).cgColor
                    button.isBordered = false
                    if let cell = button.cell as? NSButtonCell {
                        cell.isBordered = false
                    }

                    var useTitleColor = titleColor
                    if button == button_Browse || button == button_OpenFolder || button == button_Apply || button == button_Cancel {
                        useTitleColor = NSColor(white: 0.15, alpha: 1.0)
                    }

                    let attr = NSMutableAttributedString(attributedString: button.attributedTitle)
                    let range = NSRange(location: 0, length: attr.length)
                    attr.addAttribute(.foregroundColor, value: useTitleColor, range: range)
                    button.attributedTitle = attr
                }
            }
            applyStyle(toButtonsIn: sub, bgWhite: bgWhite, titleColor: titleColor)
        }
    }
}
